lib/KalturaFTPServer.js

```diff
@@ -146,28 +160,30 @@ FtpSession.prototype = {
 					}
 				}
 				else {
-					this.getFileAndInsertToCache(callback, extension, file, service, id);
+					this.getFileAndInsertToCache(callback, this, extension, file, service, id);
 				}
 			}
 		}
 	},
 
-	handleXmlListResponse: function (listResponse, callback, This, service, currentList, map, pager, currentClient, handler, filter) {
+	handleXmlListResponse: function (listResponse, callback, This, service, currentList, map, pager, currentClient, handler, filter, retrieveOnePageOnly) {
 		var document = new xmldoc(listResponse);
 		var error = document.childNamed("result").childNamed("error");
 		if (error != null) {
 			This.logErrorMsg(error.toString());
 			callback(error.toString(), null);
 		} else {
 			var objects = document.childNamed("result").childNamed("objects");
+			var count = 0;
 			objects.eachChild(function (item) {
 				This.lists[service].push(item);
 				currentList.push(item.valueWithPath("id") + '.' + map.format);
 				var itemId = item.valueWithPath("id") + '.' + map.format;
 				This.logMsg('Adding item to cache item key ' + itemId);
 				parsedList[itemId] = item;
+				count++;
 			});
-			if (objects.length == pager.pageSize) {
+			if (!retrieveOnePageOnly && objects.length == pager.pageSize && pager.pageIndex <= maxPageIndex){
```

Comment:
I think that calling the next page could happen in the caller method callback, same as the caching of each child.
These actions are also the same for JSON, iCal and XML and shouldn't be implemented three times.